### PR TITLE
Add Limit to BinaryRatingDAO

### PR DESCRIPTION
**Pull request by tajinder-txstate**
Implements #662 working towards #663. 

- Added tests for BinaryRatingDAO limiting
- Implemented BinaryRaitngDAO limiting
- Added limiting to BinaryIndexTable

### CODE REVIEW

File: lenskit-core/src/main/java/org/grouplens/lenskit/data/dao/packed/BinaryRatingDAO.java [hunk 1]
```diff
@@ -111,7 +118,7 @@ static BinaryRatingDAO fromBuffer(ByteBuffer buffer) {
         BinaryIndexTable utbl = BinaryIndexTable.fromBuffer(header.getUserCount(), tableBuffer);
         BinaryIndexTable itbl = BinaryIndexTable.fromBuffer(header.getItemCount(), tableBuffer);
 
-        return new BinaryRatingDAO(null, header, data, utbl, itbl);
+        return new BinaryRatingDAO(null, header, data, utbl, itbl, header.getRatingCount(),Long.MAX_VALUE);
```
Comment (elehack) [inline lenskit-core/src/main/java/org/grouplens/lenskit/data/dao/packed/BinaryRatingDAO.java]: Perfect.

File: lenskit-core/src/main/java/org/grouplens/lenskit/data/dao/packed/BinaryRatingDAO.java [hunk 2]
```diff
@@ -126,23 +133,49 @@ public static BinaryRatingDAO open(File file) throws IOException {
...
+            return new BinaryRatingDAO(file, header, data, utbl, itbl,header.getRatingCount(),Long.MAX_VALUE);
         } finally {
             input.close();
         }
     }
+    public  BinaryRatingDAO createWindowedView(long timestamp) {
+
+        if(timestamp>=limitTimestamp)
+        {
+            return new BinaryRatingDAO(null,header,ratingData,userTable,itemTable,limitIndex,limitTimestamp);
```
Comment (elehack) [inline lenskit-core/src/main/java/org/grouplens/lenskit/data/dao/packed/BinaryRatingDAO.java]: Can just `return this`.

Comment (elehack) [inline lenskit-core/src/main/java/org/grouplens/lenskit/data/dao/packed/BinaryRatingDAO.java]: Also, please put the opening brace on the same line as the `if`.

File: lenskit-core/src/main/java/org/grouplens/lenskit/data/dao/packed/BinaryRatingDAO.java [hunk 3]
```diff
@@ -180,20 +213,20 @@ private BinaryRatingList getRatingList(IntList indexes) {
...
-            cursor = Cursors.concat(Iterables.transform(userTable.entries(),
-                                                        new EntryToCursorTransformer()));
-            break;
-        case ITEM:
-            cursor = Cursors.concat(Iterables.transform(itemTable.entries(),
-                                                        new EntryToCursorTransformer()));
-            break;
-        default:
-            throw new IllegalArgumentException("unexpected sort order");
+            case ANY:
```
Comment (elehack) [inline lenskit-core/src/main/java/org/grouplens/lenskit/data/dao/packed/BinaryRatingDAO.java]: We need to get your IntelliJ code format settings fixed.

**Comment by elehack** [comment]
The code looks basically correct. However, on the Java 7 Travis build, there was a buffer underflow for some reason.

File: lenskit-core/src/main/java/org/grouplens/lenskit/data/dao/packed/BinaryRatingDAO.java (hunk 2 above)
Comment (tajinder-txstate) [inline lenskit-core/src/main/java/org/grouplens/lenskit/data/dao/packed/BinaryRatingDAO.java]: OK, I'll fix these issues.

File: lenskit-core/src/main/java/org/grouplens/lenskit/data/dao/packed/BinaryRatingDAO.java (hunk 3 above)
Comment (tajinder-txstate) [inline lenskit-core/src/main/java/org/grouplens/lenskit/data/dao/packed/BinaryRatingDAO.java]: I have used default formatter of IntelliJ.

File: lenskit-core/src/main/java/org/grouplens/lenskit/data/dao/packed/BinaryRatingDAO.java [hunk 4]
```diff
@@ -138,14 +145,39 @@ public static BinaryRatingDAO open(File file) throws IOException {
...
+        }
+
+        Rating r = new RatingBuilder().setUserId(0).setItemId(0).setRating(0).setTimestamp(timestamp).build();
+
+        int idx = Collections.binarySearch(getRatingList(), r, Events.TIMESTAMP_COMPARATOR);
+
+        if (idx < 0) {
+            idx = -idx - 1;
+        } else {
+            while (getRatingList().get(idx).getTimestamp() >= timestamp) {
```
Comment (elehack) [inline lenskit-core/src/main/java/org/grouplens/lenskit/data/dao/packed/BinaryRatingDAO.java]: Calling `getRatingList()` that many times creates a bunch of extra objects. Call it near the beginning of this method.

**Comment by elehack** [comment]
I'm going to go ahead and merge. We do need to make some performance improvements to this.

**Pull request merged**